src/common/src/hash/consistent_hash/mapping.rs

```diff
@@ -0,0 +1,414 @@
+// TODO: find a better place for this.
+pub type ActorId = u32;
+
+/// Trait for items that can be used as keys in [`VnodeMapping`].
+pub trait VnodeMappingItem {
+    /// The type of the item.
+    ///
+    /// Currently, there are two types of items: [`ParallelUnitId`] and [`ActorId`]. We don't use
+    /// them directly as the generic parameter because they're the same type aliases.
+    type Item: Copy + Ord + Hash + Debug;
+}
+
+/// Exapnded mapping from virtual nodes to items, essentially a vector of items and can be indexed
+/// by virtual nodes.
+pub type ExpandedMapping<T> = Vec<<T as VnodeMappingItem>::Item>;
+
+/// Generic mapping from virtual nodes to items.
+///
+/// The representation is compressed as described in [`compress_data`], which is optimized for the
+/// mapping with a small number of items and good locality.
+#[derive(Derivative)]
+#[derivative(Debug, Clone, PartialEq, Eq)]
+pub struct VnodeMapping<T: VnodeMappingItem> {
+    original_indices: Vec<u32>,
+    data: Vec<T::Item>,
+}
+
+#[expect(
+    clippy::len_without_is_empty,
+    reason = "empty vnode mapping makes no sense"
+)]
+impl<T: VnodeMappingItem> VnodeMapping<T> {
+    /// Create a uniform vnode mapping with a **set** of items.
+    ///
+    /// For example, if `items` is `[0, 1, 2]`, and the total vnode count is 10, we'll generate
+    /// mapping like `[0, 0, 0, 0, 1, 1, 1, 2, 2, 2]`.
+    pub fn new_uniform(items: impl ExactSizeIterator<Item = T::Item>) -> Self {
+        // If the number of items is greater than the total vnode count, no vnode will be mapped to
+        // some items and the mapping will be invalid.
+        assert!(items.len() <= VirtualNode::COUNT);
+
+        let mut original_indices = Vec::with_capacity(items.len());
+        let mut data = Vec::with_capacity(items.len());
+
+        let hash_shard_size = VirtualNode::COUNT / items.len();
+        let mut one_more_count = VirtualNode::COUNT % items.len();
+        let mut init_bound = 0;
+
+        for item in items {
+            let vnode_count = if one_more_count > 0 {
+                one_more_count -= 1;
+                hash_shard_size + 1
+            } else {
+                hash_shard_size
+            };
+            init_bound += vnode_count;
+
+            original_indices.push(init_bound as u32 - 1);
+            data.push(item);
+        }
+
+        // Assert that there's no duplicated items.
+        debug_assert_eq!(data.iter().duplicates().count(), 0);
+
+        Self {
+            original_indices,
+            data,
+        }
+    }
+
+    /// Create a vnode mapping where all vnodes are mapped to the same single item.
+    pub fn new_single(item: T::Item) -> Self {
+        Self::new_uniform(std::iter::once(item))
+    }
+
+    /// The length of the vnode in this mapping, typically [`VirtualNode::COUNT`].
+    pub fn len(&self) -> usize {
+        self.original_indices
+            .last()
+            .map(|&i| i as usize + 1)
+            .unwrap_or(0)
+    }
```

Comment:
Do we have an asserting semantic here? (i.e. it should always be `VirtualNode::COUNT`)

Reply:
I want to make this struct general for all lengths in case of variable vnode count in the future. But actually, it'll always be `VirtualNode::COUNT` for the uniform ones. This is covered by a unit test.

https://github.com/risingwavelabs/risingwave/blob/4731c1b00a4a54b5bde01aba1c0a04bbf83bc7a6/src/common/src/hash/consistent_hash/mapping.rs#L352-L354